### PR TITLE
429 fix implemented using watcher 

### PR DESCRIPTION
**Pull request by Bemhreth**

### CODE REVIEW

File: user_data/strategies/_429_watcher.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import watchdog.events
+import watchdog.observers
+import os
+from pathlib import Path
+
+from user_data.strategies.notifier import post_request
+# from user_data.strategies._429_file_util import delete_429_file
+
+
+class _429_Watcher(watchdog.events.PatternMatchingEventHandler):
```
Comment (tef-github) [inline user_data/strategies/_429_watcher.py]: this class still exists in execution repo, please remove from execution repo

**Review by tef-github — CHANGES_REQUESTED** [review]
changes requested

File: user_data/strategies/Strategy004.py [hunk 2]
```diff
@@ -9,15 +9,16 @@
...
+import threading
+from user_data.strategies.util import back_test, execute, create_429_watcher
 from user_data.strategies.config import Config
 from user_data.strategies.notifier import send_start_deliminator_message
 
 
 class Strategy004(IStrategy):
     if Config.IS_BACKTEST:
         send_start_deliminator_message('Freq Strategy004 ',Config.BACKTEST_COIN ,Config.BACKTEST_MONTH_LIST[Config.BACKTEST_DATA_CLEANER_MONTH_INDEX], Config.BACKTEST_DATA_CLEANER_YEAR, Config.BACKTEST_DUP, Config.BACKTEST_MAX_COUNT_DUP)
+        threading.Thread(target=create_429_watcher, args=(True,)).start()
```
Comment (tef-github) [inline user_data/strategies/Strategy004.py]: code missing in 002 and 003

File: user_data/strategies/Strategy002.py [hunk 3]
```diff
@@ -1,26 +1,32 @@
...
 
 
 class Strategy002(IStrategy):
+    controller = StrategyController()
+
     if Config.IS_BACKTEST:
-        send_start_deliminator_message('Freq Strategy002 ',Config.BACKTEST_COIN ,Config.BACKTEST_MONTH_LIST[Config.BACKTEST_DATA_CLEANER_MONTH_INDEX], Config.BACKTEST_DATA_CLEANER_YEAR, Config.BACKTEST_DUP, Config.BACKTEST_MAX_COUNT_DUP)
+        send_start_deliminator_message('Freq Strategy002 ', Config.BACKTEST_COIN, Config.BACKTEST_MONTH_LIST[Config.BACKTEST_DATA_CLEANER_MONTH_INDEX], Config.BACKTEST_DATA_CLEANER_YEAR, Config.BACKTEST_DUP, Config.BACKTEST_MAX_COUNT_DUP)
+        create_429_directory()
+        threading.Thread(target=create_429_watcher, args=(True,)).start()
```
Comment (tef-github) [inline user_data/strategies/Strategy002.py]: this can be a method in util 
_create_429_watcher_thread_

File: user_data/strategies/Strategy002.py [hunk 4]
```diff
@@ -165,15 +172,54 @@ def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: f
...
-            back_test(current_time, coin, brain)
-        elif Config.IS_EXECUTION:
-            execute(mode, coin, brain)
-        
+            self.controller.back_test(current_time, coin, brain)
+        else:
+            self.controller.execute(mode, coin, brain)
         return True
 
+    def confirm_trade_exit(self, pair: str, trade: Trade, order_type: str, amount: float,
```
Comment (tef-github) [inline user_data/strategies/Strategy002.py:186]: this already exists in develop, please merge develop in to your branch
we shouldn't see this change on the pr

File: wao/_429_watcher.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+from pathlib import Path
+
+from user_data.wao.notifier import post_request
+
+
+class _429_Watcher(watchdog.events.PatternMatchingEventHandler):
+
+    def __init__(self, is_test_mode):
+        self.is_test_mode = is_test_mode
+        # Set the patterns for PatternMatchingEventHandler
```
Comment (tef-github) [inline wao/_429_watcher.py]: remove redundant comment

File: wao/config.py [hunk 6]
```diff
@@ -1,15 +1,16 @@
...
+    BACKTEST_DUP = '0.45'
     BACKTEST_MAX_COUNT_DUP = '3'
     BACKTEST_DOWNLOADED_JSON_DATA_FILE_PATH = ""
-    BACKTEST_COIN = 'BTC'
+    BACKTEST_COIN = 'ETH'
     BACKTEST_MONTH_LIST = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']
     BACKTEST_DATA_CLEANER_YEAR = 2021
     BACKTEST_DATA_CLEANER_MONTH_INDEX = 4
-    IS_BACKTEST = False
+    IS_BACKTEST = True
```
Comment (tef-github) [inline wao/config.py]: revert

File: wao/notifier.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    post_request(text)
+
+
+def post_request(text, is_from_429_watcher=False):
+    print("post_request: " + text)
+    result = requests.post('https://api.telegram.org/bot' + executionConfig.NOTIFIER_TELEGRAM_BOT_API_TOKEN_429 +
+                           '/sendMessage?chat_id=' + executionConfig.NOTIFIER_TELEGRAM_CHANNEL_ID_BACKTEST + '&text=' + text + '&parse_mode=Markdown')
+    print(str(result))
+    if str(result) != "<Response [200]>":
```
Comment (tef-github) [inline wao/notifier.py]: can we use the specific 429 return code here

File: wao/notifier.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        delete_429_file(text)
+        write_to_429_file(text)
+    elif str(result) == "<Response [200]>" and is_from_429_watcher:
```
Comment (tef-github) [inline wao/notifier.py:28]: let's log the else case - if the error is a non-429 error

File: wao/strategy_controller.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+from wao.util import _perform_execute, _perform_back_test
+import threading
+from wao.config import Config
+import time
+
+
+class StrategyController:
```
Comment (tef-github) [inline wao/strategy_controller.py:7]: this shoudn't be visible, please merge develop to your branch

**Review by tef-github — CHANGES_REQUESTED** [review]
changes requested

File: wao/notifier.py (hunk 7 above)
Comment (Bemhreth) [reply wao/notifier.py]: 429 wont happens here but there are others which can occur when the messages overlap that is why I used this here

File: wao/strategy_controller.py (hunk 9 above)
Comment (Bemhreth) [reply wao/strategy_controller.py:7]: I did merge it but still the same

File: user_data/strategies/Strategy002.py (hunk 4 above)
Comment (Bemhreth) [reply user_data/strategies/Strategy002.py:186]: I did merge it but still the same

Comment (tef-github) [reply user_data/strategies/Strategy002.py:186]: doesn't make sense, something is wrong in the conflict resolution - create a pr with only your changes 

an existing code can not be an added change in git

File: wao/strategy_controller.py (hunk 9 above)
Comment (tef-github) [reply wao/strategy_controller.py:7]: doesn't make sense, something is wrong in the conflict resolution - **create a new pr with only your changes** 

an already existing file can not be an added change in git

Comment (Bemhreth) [reply wao/strategy_controller.py:7]: I went to the repo in github.com and copied and pasted it exactly the way it is but I still got the same thing

Comment (tef-github) [reply wao/strategy_controller.py:7]: doesn't make sense, something is wrong in the conflict resolution - **create a new pr with only your changes**

an already existing file can not be an added change in git

File: wao/util.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
+EXECUTION_PATH = '/root/workspace2/execution'  # do not move this to config
+from wao.config import Config
+import sys
+import time
+import watchdog
+import threading
```
Comment (tef-github) [inline wao/util.py:6]: this file already exists in the repo - something is wrong in the conflict resolution - create a new pr

**Pull request closed without merging**